File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Java/4ead4f11eb4579635e014b9fb04a2843.java
import java.io.*;
import java.util.*;

class DSum {
    static class Solver {
        private final BufferedReader in;
        private final PrintWriter out;

        public Solver(BufferedReader in, PrintWriter out) {
            this.in = in;
            this.out = out;
        }

        private void rd(int[] arr) throws IOException {
            StringTokenizer st = new StringTokenizer(in.readLine());
            for (int i = 0; i < arr.length; i++) {
                arr[i] = Integer.parseInt(st.nextToken());
            }
        }

        private void pr(Object... args) {
            for (Object obj : args) {
                out.print(obj);
            }
        }

        private void prln(Object... args) {
            pr(args);
            out.println();
        }

        private long rec(int l, int r, long[][] dps, long[] sums, List<List<Integer>> arrs, int k) {
            long ans = 0;
            List<Integer> carr = arrs.get(l);
            long[] dp = dps[l];
            if (l + 1 == r) {
                ans = Math.max(ans, dp[k]);
                long sum = 0;
                for (int i = 0; i < Math.min(k, carr.size()); i++) {
                    sum += carr.get(i);
                    ans = Math.max(ans, dp[k - i - 1] + sum);
                }
            } else {
                int m = (l + r) / 2;
                System.arraycopy(dps[l], 0, dps[l + 1], 0, k + 1);
                for (int i = l; i < m; i++) {
                    add_to_knap(i, dps[l + 1], k, carr, sums);
                }
                ans = Math.max(ans, rec(m, r, dps, sums, arrs, k));
                System.arraycopy(dps[l], 0, dps[l + 1], 0, k + 1);
                for (int i = m; i < r; i++) {
                    add_to_knap(i, dps[l + 1], k, carr, sums);
                }
                ans = Math.max(ans, rec(l, m, dps, sums, arrs, k));
            }
            return ans;
        }

        private void add_to_knap(int i, long[] dp, int k, List<Integer> carr, long[] sums) {
            long v = sums[i];
            int k1 = arrs.get(i).size();
            for (int j = k; j >= k1; j--) {
                dp[j] = Math.max(dp[j], dp[j - k1] + v);
            }
        }

        public void solve() throws IOException {
            StringTokenizer st = new StringTokenizer(in.readLine());
            int n = Integer.parseInt(st.nextToken());
            int k = Integer.parseInt(st.nextToken());

            long[] sums = new long[n];
            List<List<Integer>> arrs = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                st = new StringTokenizer(in.readLine());
                int t = Integer.parseInt(st.nextToken());
                List<Integer> arr = new ArrayList<>();
                for (int j = 0; j < t; j++) {
                    arr.add(Integer.parseInt(st.nextToken()));
                }
                if (t > k) {
                    arr = arr.subList(0, k);
                }
                sums[i] = arr.stream().mapToLong(Integer::longValue).sum();
                arrs.add(arr);
            }

            long[][] dps = new long[13][k + 1];
            prln(rec(0, n, dps, sums, arrs, k));
        }
    }

    public static void solve(BufferedReader in, PrintWriter out) throws IOException {
        Solver solver = new Solver(in, out);
        solver.solve();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
        solve(in, out);
        out.close();
    }
}

File: output/GPT4o-Mini/xCodeEval_4000/C++/C/4ead4f11eb4579635e014b9fb04a2843.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

typedef long long ll;

ll sum(const ll* arr, int size) {
    ll total = 0;
    for (int i = 0; i < size; i++) {
        total += arr[i];
    }
    return total;
}

void add_to_knap(ll* dp, const ll* arr, int k, ll v, int k1) {
    for (int j = k; j >= k1; j--) {
        if (dp[j] < dp[j - k1] + v) {
            dp[j] = dp[j - k1] + v;
        }
    }
}

ll rec(int l, int r, int dpi, int k, ll** dps, ll* sums, ll** arrs) {
    ll ans = 0;
    const ll* carr = arrs[l];
    ll* dp = dps[dpi];

    if (l + 1 == r) {
        ans = dp[k];
        ll sum_val = 0;
        for (int i = 0; i < (k < (int)arrs[l][0] ? k : (int)arrs[l][0]); i++) {
            sum_val += carr[i];
            if (ans < dp[k - i - 1] + sum_val) {
                ans = dp[k - i - 1] + sum_val;
            }
        }
    } else {
        int m = (l + r) / 2;
        memcpy(dps[dpi + 1], dp, sizeof(ll) * (k + 1));
        for (int i = l; i < m; i++) {
            add_to_knap(dps[dpi + 1], arrs[i], k, sums[i], arrs[i][0]);
        }
        ans = rec(m, r, dpi + 1, k, dps, sums, arrs);
        memcpy(dps[dpi + 1], dp, sizeof(ll) * (k + 1));
        for (int i = m; i < r; i++) {
            add_to_knap(dps[dpi + 1], arrs[i], k, sums[i], arrs[i][0]);
        }
        if (ans < rec(l, m, dpi + 1, k, dps, sums, arrs)) {
            ans = rec(l, m, dpi + 1, k, dps, sums, arrs);
        }
    }
    return ans;
}

void solve() {
    int n, k;
    scanf("%d %d", &n, &k);
    
    ll* sums = (ll*)malloc(n * sizeof(ll));
    ll** arrs = (ll**)malloc(n * sizeof(ll*));
    ll** dps = (ll**)malloc(13 * sizeof(ll*));
    
    for (int i = 0; i < 13; i++) {
        dps[i] = (ll*)calloc(k + 1, sizeof(ll));
    }

    for (int i = 0; i < n; i++) {
        int t;
        scanf("%d", &t);
        arrs[i] = (ll*)malloc((t + 1) * sizeof(ll));
        arrs[i][0] = t;
        for (int j = 1; j <= t; j++) {
            scanf("%lld", &arrs[i][j]);
        }
        if (t > k) {
            arrs[i][0] = k;
        }
        sums[i] = sum(arrs[i] + 1, arrs[i][0]);
    }

    printf("%lld\n", rec(0, n, 0, k, dps, sums, arrs));

    for (int i = 0; i < n; i++) {
        free(arrs[i]);
    }
    free(arrs);
    free(sums);
    for (int i = 0; i < 13; i++) {
        free(dps[i]);
    }
    free(dps);
}

int main() {
    solve();
    return 0;
}
